data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/kinesis/source/processor/KinesisRecordProcessor.java

```diff
@@ -0,0 +1,173 @@
+        this.checkpointIntervalMilliSeconds = kinesisStreamConfig.getCheckPointIntervalInMilliseconds();
+        this.buffer = buffer;
+    }
+
+    private KinesisStreamConfig getStreamConfig(final KinesisSourceConfig kinesisSourceConfig) {
+        return kinesisSourceConfig.getStreams().stream().filter(streamConfig -> streamConfig.getName().equals(streamIdentifier.streamName())).findAny().get();
+    }
+
+    @Override
+    public void initialize(InitializationInput initializationInput) {
+        // Called once when the processor is initialized.
+        kinesisShardId = initializationInput.shardId();
+        LOG.info("Initialize Processor for shard: " + kinesisShardId);
+        lastCheckpointTimeInMillis = System.currentTimeMillis();
+    }
+
+    private AcknowledgementSet createAcknowledgmentSet(final ProcessRecordsInput processRecordsInput) {
+        return acknowledgementSetManager.create((result) -> {
+            acknowledgementSetCallbackCounter.increment();
+            if (result) {
+                LOG.info("acknowledgements received");
+                checkpoint(processRecordsInput.checkpointer());
+            } else {
+                LOG.info("acknowledgements received with false");
+            }
+
+        }, ACKNOWLEDGEMENT_SET_TIMEOUT);
+    }
+
+    @Override
+    public void processRecords(ProcessRecordsInput processRecordsInput) {
+        List<Record<Event>> records = new ArrayList<>();
+
+        try {
+            Optional<AcknowledgementSet> acknowledgementSetOpt = Optional.empty();
+            boolean acknowledgementsEnabled = kinesisSourceConfig.isAcknowledgments();
+            if (acknowledgementsEnabled) {
+                acknowledgementSetOpt = Optional.of(createAcknowledgmentSet(processRecordsInput));
+            }
+
+            for (KinesisClientRecord record : processRecordsInput.records()) {
+                processRecord(record, records::add);
+            }
+
+            acknowledgementSetOpt.ifPresent(acknowledgementSet -> records.forEach(record -> acknowledgementSet.add(record.getData())));
+
+            buffer.writeAll(records, bufferTimeoutMillis);
+
+            acknowledgementSetOpt.ifPresent(AcknowledgementSet::complete);
+
+            // Checkpoint for shard
+            if (kinesisStreamConfig.isEnableCheckPoint() && System.currentTimeMillis() - lastCheckpointTimeInMillis > checkpointIntervalMilliSeconds) {
+                LOG.info("Regular checkpointing for shard " + kinesisShardId);
+                checkpoint(processRecordsInput.checkpointer());
+                lastCheckpointTimeInMillis = System.currentTimeMillis();
+            }
+        } catch (Exception ex) {
+            recordProcessingErrors.increment();
+            LOG.error("Failed writing shard data to buffer: ", ex);
+        }
+    }
+
+    private void processRecord(KinesisClientRecord record, Consumer<Record<Event>> eventConsumer) throws IOException {
```

Comment:
I think we should split the logic for processing an individual record out of this class to make the class easier to extend and test over time.

Reply:
@sb2k16 , Let's make this a separate class. Perhaps `KinesisRecordConverter`. It can be responsible for all the work for taking the record and using the codec. You can have some good tests on this as well.

We've had difficulty expanding our tests in other similar locations because we combine too many things in a single class.